### PR TITLE
linux-capture: XCompCap now chooses glXFBConfigs based on window depth

### PR DESCRIPTION
**Pull request by SingularityAzure**
Rather than inheriting the exact visuals from the window,
find whether the color buffer has alpha, find a config with
the necessary attributes for capture, and match its depth to
the depth of the window instead of the color buffer. Also,
cleanup glxpixmap texture binding before destroying the texture.

### CODE REVIEW

File: plugins/linux-capture/xcompcap-main.cpp [hunk 1]
```diff
@@ -351,18 +354,88 @@ void XCompcapMain::updateSettings(obs_data_t *settings)
...
+		GLX_BIND_TO_TEXTURE_RGBA_EXT, GL_TRUE,
+		GLX_DRAWABLE_TYPE, GLX_PIXMAP_BIT,
+		GLX_BIND_TO_TEXTURE_TARGETS_EXT, GLX_TEXTURE_2D_BIT_EXT,
+		GLX_ALPHA_SIZE, 8,
+		GLX_DOUBLEBUFFER, GL_FALSE,
+		None
+	};
+
+	int nelem = 0;
+	GLXFBConfig* configs = glXGetFBConfigs(xdisp,
```
Comment (RytoEX) [inline plugins/linux-capture/xcompcap-main.cpp]: Style:  The asterisk should be next to `configs`.

File: plugins/linux-capture/xcompcap-main.cpp [hunk 2]
```diff
@@ -351,18 +354,88 @@ void XCompcapMain::updateSettings(obs_data_t *settings)
...
+		p->height = 0;
+		p->width = 0;
+		return;
+	}
+
+	GLXFBConfig config;
+	for (int i = 0; i < nelem; i++) {
+		config = configs[i];
+		XVisualInfo *visual = glXGetVisualFromFBConfig(xdisp, config);
+		if (!visual) {
```
Comment (RytoEX) [inline plugins/linux-capture/xcompcap-main.cpp]: Style:  This `if` statement does not need curly braces.  Leave a blank line below the `if` contents.

File: plugins/linux-capture/xcompcap-main.cpp [hunk 3]
```diff
@@ -351,18 +354,88 @@ void XCompcapMain::updateSettings(obs_data_t *settings)
...
+		}
+		if (attr.visual->visualid != visual->visualid) {
+			XFree(visual);
+			continue;
+		}
+		XFree(visual);
+
+		int value;
+		glXGetFBConfigAttrib(xdisp, config, GLX_ALPHA_SIZE, &value);
+		if (value != 8) {
```
Comment (RytoEX) [inline plugins/linux-capture/xcompcap-main.cpp]: Style:  This `if` statement does not need curly braces.  Leave a blank line below the `if` contents.

File: plugins/linux-capture/xcompcap-main.cpp [hunk 4]
```diff
@@ -351,18 +354,88 @@ void XCompcapMain::updateSettings(obs_data_t *settings)
...
+		p->win = 0;
+		p->height = 0;
+		p->width = 0;
+		return;
+	}
+	bool found = false;
+	for (int i = 0; i < nelem; i++) {
+		config = configs[i];
+		XVisualInfo *visual = glXGetVisualFromFBConfig(xdisp, config);
+		if (!visual) {
```
Comment (RytoEX) [inline plugins/linux-capture/xcompcap-main.cpp]: Style:  This `if` statement does not need curly braces.  Leave a blank line below the `if` contents.

**Comment by SingularityAzure** [comment]
@RytoEX I was unaware of the conversation going on on Mantis, so thanks for bringing that to my attention. I'll look into that patch when I have some extra time. It looks like it may help fix the issues I'm still having on Arch with Chromium and Nautilus.
Also style conformance changes have been committed and squashed.

**Comment by Lucretia** [comment]
This fixes issues with recording failinging due to:

error: glCopyImageSubData failed, glGetError returned 0x502
error: device_copy_texture (GL) failed

VSCode recording now works for me.

I still get the errors if I enable "Use alpha-less texture format (mesa)"

```
$ glxinfo -B
name of display: :0
display: :0  screen: 0
direct rendering: Yes
Extended renderer info (GLX_MESA_query_renderer):
    Vendor: X.Org (0x1002)
    Device: AMD Radeon (TM) R9 380 Series (TONGA, DRM 3.27.0, 4.19.0-rc4, LLVM 7.0.0) (0x6939)
    Version: 18.3.0
    Accelerated: yes
    Video memory: 4096MB
    Unified memory: no
    Preferred profile: core (0x1)
    Max core profile version: 4.5
    Max compat profile version: 4.5
    Max GLES1 profile version: 1.1
    Max GLES[23] profile version: 3.2
Memory info (GL_ATI_meminfo):
    VBO free memory - total: 3581 MB, largest block: 3581 MB
    VBO free aux. memory - total: 4079 MB, largest block: 4079 MB
    Texture free memory - total: 3581 MB, largest block: 3581 MB
    Texture free aux. memory - total: 4079 MB, largest block: 4079 MB
    Renderbuffer free memory - total: 3581 MB, largest block: 3581 MB
    Renderbuffer free aux. memory - total: 4079 MB, largest block: 4079 MB
Memory info (GL_NVX_gpu_memory_info):
    Dedicated video memory: 4096 MB
    Total available memory: 8192 MB
    Currently available dedicated video memory: 3581 MB
OpenGL vendor string: X.Org
OpenGL renderer string: AMD Radeon (TM) R9 380 Series (TONGA, DRM 3.27.0, 4.19.0-rc4, LLVM 7.0.0)
OpenGL core profile version string: 4.5 (Core Profile) Mesa 18.3.0-devel (git-a9475d9337)
OpenGL core profile shading language version string: 4.50
OpenGL core profile context flags: (none)
OpenGL core profile profile mask: core profile

OpenGL version string: 4.5 (Compatibility Profile) Mesa 18.3.0-devel (git-a9475d9337)
OpenGL shading language version string: 4.50
OpenGL context flags: (none)
OpenGL profile mask: compatibility profile

OpenGL ES profile version string: OpenGL ES 3.2 Mesa 18.3.0-devel (git-a9475d9337)
OpenGL ES profile shading language version string: OpenGL ES GLSL ES 3.20
```

**Comment by Rua** [comment]
This PR doesn't fix the issue for me, sadly. I'm still unable to capture the terminal, the Steam window or Skyrim. I can't capture the Desktop window either. Other windows do work.

**Comment by Rua** [comment]
With this PR, I am able to capture a terminal window, which I wasn't able to before. But I'm still not able to capture the Steam window or Skyrim.

**Comment by Lucretia** [comment]
I'm back to no capture working on Linux with VSCode and --disable-gpu doesn't work.

**Comment by jp9000** [comment]
@Rua -- are you able to capture those things on the latest stable release (not master)?

**Comment by Rua** [comment]
With the stable release 22.0.3 I'm not able to capture desktop, terminal, Steam or Skyrim. The same as with the previous stable release.

**Comment by jp9000** [comment]
@Lucretia does this PR function any differently from the current stable release?

**Comment by Lucretia** [comment]
@jp9000 Current git tree is what I've been using. I'm now using a custom build in which I comment out the glCopyImageSubData calls and fall back to FBO's. This works fine.

**Comment by Lucretia** [comment]
I also, compiled a lot of tagged versions and none of them worked. Early version = red screen, later versions = black screen. I think it's a mesa bug somewhere.

**Comment by Lucretia** [comment]
Here's the patch.

[0001-Force-FBO-use-instead-of-glCopyImageSubData-due-to-i.patch.gz](https://github.com/obsproject/obs-studio/files/2655633/0001-Force-FBO-use-instead-of-glCopyImageSubData-due-to-i.patch.gz)

**Comment by willianrampazzo** [comment]
Just to mention, with the current master as of today, I'm able to record Terminator screen but still unable to record Chrome browser.

**Comment by Lucretia** [comment]
Just to add, there should be a switch for for this to drop back to FBO's if the latest stuff isn't working, which it isn't.

**Comment by willianrampazzo** [comment]
> Just to mention, with the current master as of today, I'm able to record Terminator screen but still unable to record Chrome browser.

I tested with Firefox and it worked fine, so my problem is just with Chrome now. I tried to disable hardware acceleration without success.

File: plugins/linux-capture/xcompcap-main.cpp [hunk 5]
```diff
@@ -351,18 +354,88 @@ void XCompcapMain::updateSettings(obs_data_t *settings)
...
+			continue;
 		}
+		XFree(visual);
+		found = true;
+		break;
+	}
+	if (!found)
+		config = configs[0];
+
+	if (cf == GS_BGRX || !has_alpha) {
```
Comment (GloriousEggroll) [inline plugins/linux-capture/xcompcap-main.cpp]: changing !has_alpha to has_alpha fixes the transparency bug mentioned here:

https://github.com/obsproject/obs-studio/pull/1485

I may be wrong but from what I see, has_alpha is set to true by default, which would mean it should check if has_alpha remains true before setting opaqueness.  Here from what it seems, you're setting opaqueness even though has_alpha is false.. which doesn't seem right no?

Comment (SingularityAzure) [reply plugins/linux-capture/xcompcap-main.cpp]: My original purpose in making these changes was to enable capture of windows with Alpha, such that the alpha is used in compositing the capture (transparent windows remain transparent when captured). This is helpful for a number of reasons, such as some GTK+ windows that include the shadow in their pixmap, and (as was my main use-case) to capture a transparent overlay window I was developing for streaming purposes.
Also, when a window doesn't have alpha, it needs to be drawn opaque or else it's drawn with an alpha of 0, making it invisible.
This behavior _should_ be toggle-able for transparent windows by disabling transparency, but that appears to not work in all situations.

**Comment by SingularityAzure** [comment]
> Here's the patch.
> 
> [0001-Force-FBO-use-instead-of-glCopyImageSubData-due-to-i.patch.gz](https://github.com/obsproject/obs-studio/files/2655633/0001-Force-FBO-use-instead-of-glCopyImageSubData-due-to-i.patch.gz)

This patch breaks all xcomposite capture on my machine running Gnome on Arch.

**Comment by SuslikV** [comment]
The initial value of GL_TEXTURE_MIN_FILTER is GL_NEAREST_MIPMAP_LINEAR
The mentioned patch doesn't sets min filter for both target and source like:
`glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST)`
and don't provide any checks on this before calling copy.

**Pull request merged**